### PR TITLE
feat: create no models flag for generate command

### LINKED ISSUE #11582: Generating Prisma client without any model in its schema
## Problem

We have a server template including Prisma, which works fine so far. However, we now have another server extending this template without the need of models - just raw queries. While generating the Prisma client, via `npx prisma generate`, it is complaining about not having at least one model.

```
$ npx prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Error: 
You don't have any models defined in your schema.prisma, so nothing will be generated.
```

Obviously correct - but in this case not wanted. So, we would like to use Prisma with basically `$queryRaw` and `$queryRawUnsafe` only.

## Suggested solution

It should be allowed to generate the Prisma Client without existing models in its schema.

## Alternatives

First guess would be to add a dummy model.

### PR DESCRIPTION
**Pull request by gutyerrez**
This feature request brings a new flag to the client generation

Closes https://github.com/prisma/prisma/issues/11582

### CODE REVIEW

**Comment by gutyerrez** [comment]
> Hi @gutyerrez 👋🏼
> 
> Is there an existing related issue about this? I'm interested in the context of this PR, like why is that useful for you or others?

Hello @Jolg42!

Hasn't any issue about this yet. That's a suggestion for an case i've experienced... I can explain that:

I have N micro services and one of this doesn't need models but need's  to execute queries on database. I use prisma in all other applications. I like prisma so much and the way it's works and i don't like to choose other db tool, the best way is allowing generation of prisma without models and using it

**Comment by janpio** [comment]
Do I understand correctly that this PR suggests a new `prisma generate` flag `--allow-no-models` that allows running `prisma generate` even with no models defined in your Prisma schema, instead of getting an error message that suggests to add a model instead?

File: packages/client/src/generation/TSClient/PrismaClient.ts [hunk 1]
```diff
@@ -334,7 +334,18 @@ export class PrismaClientClass implements Generatable {
   private get jsDoc(): string {
     const { dmmf } = this
 
-    const example = dmmf.mappings.modelOperations[0]
+    let example: DMMF.ModelMapping
```
[the remark this answers is not in the document]
Comment (gutyerrez) [reply packages/client/src/generation/TSClient/PrismaClient.ts:337]: When the `generate prisma` command runs successfully a output message containing a example of Prisma is printed in this case i verify if any model exists when none i print a fake example of code usage with models

**Comment by gutyerrez** [comment]
> Do I understand correctly that this PR suggests a new `prisma generate` flag `--allow-no-models` that allows running `prisma generate` even with no models defined in your Prisma schema, instead of getting an error message that suggests to add a model instead?
> 
> 
> 
> If so, we actually do have an existing issue for this problem that this PR could resolve and close: https://github.com/prisma/prisma/issues/11582

Hello @janpio is exactly that you said

**Comment by janpio** [comment]
There is a type error left:
> Error: packages/client/src/generation/dmmf.ts(111,37): error TS2540: Cannot assign to 'prisma' because it is a read-only property.
https://github.com/prisma/prisma/pull/22729/files#diff-9fd284e77064cb32b2208c5d7e2430f75ff7cff3088597096f499d8a0b856023

**Comment by gutyerrez** [comment]
> There is a type error left:
> 
> > Error: packages/client/src/generation/dmmf.ts(111,37): error TS2540: Cannot assign to 'prisma' because it is a read-only property.
> 
> https://github.com/prisma/prisma/pull/22729/files#diff-9fd284e77064cb32b2208c5d7e2430f75ff7cff3088597096f499d8a0b856023

Dear @janpio i fixed this problem. But now i have this test failing:

https://github.com/prisma/prisma/actions/runs/7998281868/job/21844241882?pr=22729

But when i run test locally it's works:

image

**Comment by janpio** [comment]
Started rerun of test.

**Comment by janpio** [comment]
Thank you, next step then is for someone from the Engineering team to actually review and approve the code. But you are in a good position for that.

**Review by janpio — APPROVED** [review]
All good from the product side, if this works as it looks like, this fulfills what we wanted in https://github.com/prisma/prisma/issues/11582 👍

**Comment by fredriklugg** [comment]
Will this get merged soon? 🤔

**Review by Druue — CHANGES_REQUESTED** [review]
Hey @gutyerrez,

This generally looks good, a couple things however:
- Can you please add some tests for this; We have the following two tests already in place:
    - [fail if no model(s) found - sqlite](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/packages/internals/src/__tests__/getGenerators/getGenerators.test.ts#L720-L756)
    - [fail if no model(s) found - mongodb](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/packages/internals/src/__tests__/getGenerators/getGenerators.test.ts#L758-L794)
    
    So this would be a good place to start
    
- I took a look locally and ran into the following error in a [fresh sandbox](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/CONTRIBUTING.md#setting-up-a-locally-linked-development-folder)
    ```
    pnpm prisma generate --allow-no-models 
    Prisma schema loaded from prisma/schema.prisma
    {}
    {}
    Error: 
    Cannot read properties of undefined (reading 'reduce')
    ```
    This was using the following schema
    ```prisma
    generator client {
        provider = "prisma-client-js"
        output   = "../node_modules/.prisma/client"
    }

    datasource db {
        provider = "sqlite"
        url      = "file:./dev.db"
    }
    ```

**Comment by Druue** [comment]
Hey @gutyerrez, we are going to close this PR for now. If you ever get some time to look into the aforementioned errors, adding the tests, and showing that both engines and client work without models, please feel free to open a new one ✨

**Comment by gutyerrez** [comment]
> Hey @gutyerrez,
> 
> This generally looks good, a couple things however:
> 
> * Can you please add some tests for this; We have the following two tests already in place:
>   
>   * [fail if no model(s) found - sqlite](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/packages/internals/src/__tests__/getGenerators/getGenerators.test.ts#L720-L756)
>   * [fail if no model(s) found - mongodb](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/packages/internals/src/__tests__/getGenerators/getGenerators.test.ts#L758-L794)
>   
>   So this would be a good place to start
> * I took a look locally and ran into the following error in a [fresh sandbox](https://github.com/prisma/prisma/blob/6542fb22fb1ff449968a2a6144afd53f1807e6ac/CONTRIBUTING.md#setting-up-a-locally-linked-development-folder)
>   ```
>   pnpm prisma generate --allow-no-models 
>   Prisma schema loaded from prisma/schema.prisma
>   {}
>   {}
>   Error: 
>   Cannot read properties of undefined (reading 'reduce')
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   This was using the following schema
>   ```prisma
>   generator client {
>       provider = "prisma-client-js"
>       output   = "../node_modules/.prisma/client"
>   }
>   
>   datasource db {
>       provider = "sqlite"
>       url      = "file:./dev.db"
>   }
>   ```

Thanks, @Druue i'll fix this problem and reopen PR later

**Pull request closed without merging**